autorest/azure/auth/auth.go

```diff
@@ -41,59 +41,105 @@ import (
 // 3. Username password
 // 4. MSI
 func NewAuthorizerFromEnvironment() (autorest.Authorizer, error) {
-	tenantID := os.Getenv("AZURE_TENANT_ID")
-	clientID := os.Getenv("AZURE_CLIENT_ID")
-	clientSecret := os.Getenv("AZURE_CLIENT_SECRET")
-	certificatePath := os.Getenv("AZURE_CERTIFICATE_PATH")
-	certificatePassword := os.Getenv("AZURE_CERTIFICATE_PASSWORD")
-	username := os.Getenv("AZURE_USERNAME")
-	password := os.Getenv("AZURE_PASSWORD")
-	envName := os.Getenv("AZURE_ENVIRONMENT")
-	resource := os.Getenv("AZURE_AD_RESOURCE")
-
-	var env azure.Environment
+	settings := getEnvVars()
+
+	env, err := getEnv(settings.envName)
+	if err != nil {
+		return nil, err
+	}
+
+	if settings.resource == "" {
+		settings.resource = env.ResourceManagerEndpoint
+	}
+
+	return getAuth(settings, env)
+}
+
+// NewKeyvaultAuthorizerFromEnvironment creates a keyvault dataplane Authorizer configured from environment variables in the order:
+// 1. Client credentials
+// 2. Client certificate
+// 3. Username password
+// 4. MSI
+func NewKeyvaultAuthorizerFromEnvironment() (autorest.Authorizer, error) {
```

Comment:
I see the point in case you need to auth to multiple resources . but does it make more sense to have a `NewAuthorizerFromEnvironmentWithResource` instead of creating a seperate function for each ? That way whatever resource they have they have they can pass it without making a request.

Reply:
What if we had both, with the specialized ones building on the general ones?  E.g. look at line 73, we see customers use env.KeyVaultEndpoint as the resource for key vault which fails due to trailing '/' character.  I'm sure there are other solutions to that particular issue, but I think there's value in having "brain-dead simple" authorizers.

Reply:
To add to my previous comment, if we think it makes more sense to have these specialized ones as convenience methods in the SDK (i.e. not go-autorest) it works for me.

Reply:
Fixed

Reply:
I opened a PR on the SDK with Joel's suggestion